README.md

```diff
@@ -1,526 +1,917 @@
+  context.defineMetadata(MY_META, true);
+}
+
+@myMeta
+class C {
+  @myMeta foo = 123
+  @myMeta bar() {}
+
+  @myMeta static baz = 123;
+  @myMeta static qux() {}
 }
 ```
 
-This decorator could be defined as follows:
+Would produce the following two metadata objects:
 
 ```js
-const handler = Symbol("handler");
-function on(eventName)
-  return (method, context) => {
-    context.metadata = {[handler]: eventName};
-    return method;
-  }
+let staticMeta = {
+  constructor: { "my-meta": true, [MY_META]: true },
+  baz: { "my-meta": true, [MY_META]: true },
+  qux: { "my-meta": true, [MY_META]: true },
 }
 
-class MetadataLookupCache {
-  #map = new WeakMap();
-  #name;
-  constructor(name) { this.#name = name; }
-  get(newTarget) {
-    let data = this.#map.get(newTarget);
-    if (data === undefined) {
-      data = [];
-      let klass = newTarget;
-      while (klass !== null && !(this.#name in klass)) {
-        for (const [name, {[this.#name]: eventName}]
-             of Object.entries(klass[Symbol.metadata].instance.methods)) {
-          if (eventName !== undefined) {
-            data.push({name, eventName});
-          }
-        }
-        klass = klass.__proto__;
-      }
-      this.#map.set(newTarget, data)
-    }
-    return data;
-  }
+let nonStaticMeta = {
+  foo: { "my-meta": true, [MY_META]: true },
+  bar: { "my-meta": true, [MY_META]: true },
 }
+```
 
-let handlersMap = new MetadataLookupCache(handler);
+Note that metadata defined by a class decorator is assigned to the static `constructor` key. This is because `constructor` is a reserved name within class definitions, so it cannot conflict with another class element with the same name.
 
-function WithActions(superclass) {
-  return class C extends superclass {
-    constructor(...args) {
-      super(...args);
-      let handlers = handlersMap.get(new.target, C);
-      for (const {name, eventName} of handlers) {
-        this.addEventListener(eventName, this[name].bind(this));
-      }
-    }
-  }
+These metadata objects are then exposed via the `Symbol.metadata` property on the class (for static metadata) and the class prototype (for non-static metadata). So the above example is roughly equivalent, when executed, to:
+
+```js
+C[Symbol.metadata] = {
+  constructor: { "my-meta": true, [MY_META]: true },
+  baz: { "my-meta": true, [MY_META]: true },
+  qux: { "my-meta": true, [MY_META]: true },
+};
+
+C.prototype[Symbol.metadata] = {
+  foo: { "my-meta": true, [MY_META]: true },
+  bar: { "my-meta": true, [MY_META]: true },
+};
+```
+
+This is not quite a "desugaring", since we would still need to execute the decorators to determine what metadata would be generated. This is mainly for illustrative purposes.
+
+Private class elements can also expose metadata via this API. When this occurs, the metadata is exposed on a key which has the same spelling as the private element in code:
+
+```js
+class C {
+  @myMeta #x;
 }
+
+C.prototoype[Symbol.metadata]["#x"];
```

Comment:
What happens if i also have `@myMeta ['#x'] = whatever;` on the same class?

Reply:
Ah, this is a great point, I hadn't considered that. I think we'll need to change this somehow to account for that.